### ISSUE TITLE
[feature request] Option to make project directory as a GOPATH

### ISSUE THREAD

**Issue by akamensky** [go-lang-plugin-org/go-lang-idea-plugin#2142]
Preamble:

I know that general recommendation for GO is to have some global permanent location for modules to be downloaded and installed to. However I find it as a wrong approach in terms of go. 

The reason is because other languages/package managers (and good example here is Maven for Java) that use this way provide the ability to have dependency on version of package and same time can download and keep multiple versions of same package.

`go get/install` does not provide version dependencies nor provides structured approach to this issue.

So as a GO developer I am used to install dependencies locally defining `$GOPATH` for each project separately, thus doing exactly as title says. However setting this for my Idea each time I open another project is a hell of a workaround. As of now I always have my imports starting with './' so that auto-completion and deps loading for running/debugging would work.

Thus after all above I would like to see the option for all go projects to have GOPATH to be set as project directory. 

This would simultaneously resolve issue of always seeing `GOPATH is empty` message that many people said about already.

**Comment by zolotov** [comment]
Just add project directory into Project Libraries in configuration, what's the problem?

**Comment by dlsniper** [comment]
> I know that general recommendation for GO is to have some global permanent location for modules to be downloaded and installed to. However I find it as a wrong approach in terms of go.

You are fighting the language in the wrong place, please see https://groups.google.com/forum/#!forum/golang-nuts for a better place to discuss this issue.

Also this will change with the Go vendoring solution (which is not yet supported by the plugin).

> As of now I always have my imports starting with './' so that auto-completion and deps loading for running/debugging would work.

Also incorrect and wrong due to other reasons, search on the link above.

If you define a GOPATH per project then the way to solve this in IDEA (and in general) is to make sure your GOPATH (aka project root dir) respects the https://golang.org/doc/code.html#Workspaces definition. Everything will work as intended afterwards, no need for workarounds (not even setting anything in the Go Libraries section).

As for the message that the GOPATH is empty, @zolotov do you think we should add a dismiss button to it? In a scenario where the project contains the GOPATH this can become redundant.

**Comment by zolotov** [comment]
> As for the message that the GOPATH is empty, @zolotov do you think we should add a dismiss button to it? In a scenario where the project contains the GOPATH this can become redundant.

No I don't see how it can be redundant.

**Comment by dlsniper** [comment]
> No I don't see how it can be redundant.

Sorry, misinformation. I thought we automatically add the project dir to the GOPATH but we don't do it.

> , no need for workarounds (not even setting anything in the Go Libraries section).

Last part is not correct, you still need to add the project dir to Settings -> Language & Frameworks -> Go -> Go Libraries -> Project libraries section.

**Comment by akamensky** [comment]
Would it worth adding project directory to the GOPATH by default?

**Comment by dlsniper** [comment]
No, sorry, it's not applicable to most cases, even more so with Go 1.6+ (the vendor experiment graduates to non-experiment anymore).

I wanted to suggest to add an option in the empty GOPATH message to do that but at this point I think I'd rather not do that because it might further confuse users. @zolotov what do you think?

**Comment by akamensky** [comment]
I see, then I guess close this issue. The dismiss button is a separate issue IMO...

**Comment by zolotov** [comment]
I believe I can implement support for path-marcos (e.g. you'll be able to set up $PROJECT_DIR$ as go library path)
